providers/repo.rb

```diff
@@ -138,17 +138,45 @@ def install_gem
   end
 end
 
-def read_token(repo_url, gems=false)
-  return repo_url unless new_resource.master_token
+### The hostname lookup chain is:
+### 1) node['packagecloud']['override_hostname']
+### 2) node['fqdn']
+### 3) node['hostname']
+### We persist this hostname result in node['packagecloud']['_stored_hostname']
+### between chef runs so that we can determine if a new read token should be issued
+### from the packagecloud API for new/changed hostnames. See #should_issue_read_token?
```

Comment:
I think you're working too hard to persist the result.  Except under edge cases, the hostname selected by the algorithm should be the identical for every convergence.  This is especially true if the user defined `node['packagecloud']['hostname']` in the node JSON data (either directly or via roles or environments).